docs/proposals/llm-hyperparameter-optimization-api.md

```diff
@@ -0,0 +1,235 @@
+  * [Links](#links)
+  * [Motivation](#motivation)
+  * [Goals](#goals)
+  * [Non-Goals](#non-goals)
+  * [Design for API](#design-for-api)
+    + [Example](#example)
+  * [Implementation](#implementation)
+
+## Links
+
+- [katib/issues#2291 (Tuning API in Katib for LLMs)](https://github.com/kubeflow/katib/issues/2291)
+
+## Motivation
+
+The rapid advancements and growing popularity of Large Language Models (LLMs) have driven an increased need for effective LLMOps in Kubernetes environments. To address this, we developed a [train API](https://www.kubeflow.org/docs/components/training/user-guides/fine-tuning/) within the Training Python SDK, simplifying the process of fine-tuning LLMs using distributed PyTorchJob workers. However, hyperparameter optimization remains a crucial yet labor-intensive task for enhancing model performance. Automating this tuning process through a dedicated API will facilitate efficient and scalable exploration of hyperparameters, ultimately improving model performance and reducing manual effort.
+
+## Goals
+
+Our goal is to develop a high-level API for tuning hyperparameters of LLMs that simplifies the process of hyperparameter optimization in Kubernetes. This API will seamlessly integrate with external platforms like HuggingFace and S3 for importing pretrained models and datasets. By specifying parameters for the training objective, trial configurations, and PyTorch worker configurations, the API will automate the creation of experiments and execution of trials. This abstraction of Kubernetes infrastructure complexities will enable data scientists to optimize hyperparameters efficiently and effectively.
+
+## Non-Goals
+
+1. Incorporate early stopping strategy into the API to optimize training efficiency.
+2. Expand support for distributed training in frameworks beyond PyTorch by leveraging their distributed training capabilities.
+3. Support adding custom providers through configmap or CRD approach to enhance flexibility.
+4. Enable users to deploy tuned models for inference within their applications or seamlessly integrate them into existing NLP pipelines for specialized tasks.
+
+## Design for API
+
+![Design for API](images/design_api.jpg)
+
+```python
+import kubeflow.katib as katib
+from kubeflow.katib import KatibClient
+
+class KatibClient(object):
+	
+	def tune(
+		name: str, 
+		namespace: Optional[str] = None,
+		model_provider_parameters = None,
+		dataset_provider_parameters = None,
+		trainer_parameters = None,
+		algorithm_name: str = "random",
+		algorithm_settings: Union[dict, List[models.V1beta1AlgorithmSetting], None] = None,
+		objective_metric_name: str = "eval_accuracy",
+		additional_metric_names: List[str] = [],
+		objective_type: str = "maximize",
+		objective_goal: float = None,
+		max_trial_count: int = None,
+		parallel_trial_count: int = None,
+		max_failed_trial_count: int = None,
+		resources_per_trial = models.DistributedResources(
```

Comment:
How do you like the name `TrainerConfig` to be consistent with API for TrainJob in Training Operator V2 APIs?
https://docs.google.com/document/d/1bha8rB6_iPTi9rXjJMnKi-CLxfL7dwtmQCKH4N6dcsI/edit#heading=h.xmn2xt4wqxyc

Reply:
@andreyvelich Do you think `TrialResourceSpec` would be better? It seems more consistent with the context here. Additionally, since most of the classes we have defined include "V1beta1" at the beginning, should we also incorporate it here for consistency?

Reply:
The models that are defined here: https://github.com/kubeflow/katib/tree/master/sdk/python/v1beta1/kubeflow/katib/models are autogenerated using this script: https://github.com/kubeflow/katib/blob/master/hack/gen-python-sdk/gen-sdk.sh.
For this new class, we might need to create a new SDK directory `types` and add a new class there.

---
I was thinking that if users don't need to use distributed training as Trials, they can just simple use `resources_per_trial` as follows:
```python
tune(
  resources_per_trial={
      "gpu": 2,
      "cpu" 10,
    }
)
```

Otherwise, when they want to use distributed training within Trials, users need to use special type in `resources_per_trial` as `TrainerConfig`. We can also call it `types.TrainerResources`.

Any thoughts @helenxie-bit @droctothorpe @johnugeorge @tenzen-y ?

Reply:
I think it makes sense. We can include examples for both scenarios in the proposal.

Reply:
@andreyvelich I have updated the configuration for `resources_per_trial`, which includes both options. Please review it.